File: experiments/synthetic_hydropower/clients/r/adapter.R
# Reusable persistent-worker adapter for synthetic hydropower optimisation.

new_hydropower_worker <- function(command = Sys.getenv(
  "SYNTHETIC_HYDROPOWER_COMMAND",
  "synthetic-hydropower"
)) {
  if (.Platform$OS.type == "windows" && grepl("\\.cmd$", command, ignore.case = TRUE)) {
    python <- file.path(dirname(command), "python.exe")
    script <- sub("\\.cmd$", "", command, ignore.case = TRUE)
    if (!file.exists(python) || !file.exists(script)) {
      stop("the Windows hydropower launcher is incomplete: ", command)
    }
    return(processx::process$new(
      python, args = c("-u", script, "worker"), stdin = "|", stdout = "|", stderr = "|"
    ))
  }
  processx::process$new(command, args = "worker", stdin = "|", stdout = "|", stderr = "|")
}

close_hydropower_worker <- function(worker) {
  if (worker$is_alive()) worker$kill()
  invisible(NULL)
}

project_power <- function(vector, definition) {
  values <- matrix(as.numeric(vector), nrow = 2L, byrow = TRUE)
  minimum <- as.numeric(definition$power_bounds_mw$minimum)
  maximum <- as.numeric(definition$power_bounds_mw$maximum)
  for (plant in seq_len(2L)) {
    values[plant, ] <- pmin(pmax(values[plant, ], 0), maximum[plant])
    values[plant, values[plant, ] < minimum[plant] / 2] <- 0
    values[plant, values[plant, ] > 0 & values[plant, ] < minimum[plant]] <- minimum[plant]
  }
  values
}

evaluate_hydropower <- function(worker, vector, definition, case_id = "candidate") {
  schedule <- project_power(vector, definition)
  payload <- list(
    schema_version = "synthetic-hydropower/v1",
    requests = list(list(
      case_id = case_id,
      incremental_inflow_m3s = definition$incremental_inflow_m3s,
      target_power_mw = unname(split(schedule, row(schedule)))
    ))
  )
  worker$write_input(paste0(jsonlite::toJSON(payload, auto_unbox = TRUE, digits = NA), "\n"))
  status <- worker$poll_io(30000)
  if (status[["output"]] != "ready") {
    stop("hydropower worker did not return a response")
  }
  response <- jsonlite::fromJSON(worker$read_output_lines(n = 1L), simplifyVector = FALSE)
  if (!is.null(response$error)) stop(response$error$message)
  response$results[[1L]]
}

hydropower_objective <- function(worker, vector, definition) {
  result <- evaluate_hydropower(worker, vector, definition)
  value <- result$simulation$objective
  if (!is.numeric(value) || length(value) != 1L || !is.finite(value)) {
    stop("hydropower response has no finite canonical objective")
  }
  value
}

File: experiments/synthetic_hydropower/clients/julia/adapter.jl
"""Persistent JSON-lines client for the canonical synthetic hydropower worker."""

using JSON

function worker_command()
    command = get(ENV, "SYNTHETIC_HYDROPOWER_COMMAND", "synthetic-hydropower")
    if Sys.iswindows() && endswith(lowercase(command), ".cmd")
        python = joinpath(dirname(command), "python.exe")
        script = command[1:end-4]
        isfile(python) && isfile(script) || error("incomplete Windows hydropower launcher")
        return Cmd([python, "-u", script, "worker"])
    end
    Cmd([command, "worker"])
end

mutable struct HydropowerWorker
    io::IO
end

function HydropowerWorker()
    return HydropowerWorker(open(worker_command(), "r+"))
end

function close_hydropower_worker(worker::HydropowerWorker)
    Base.close(worker.io)
end

function project_power(vector::Vector{Float64}, definition)
    periods = Int(definition["periods"])
    length(vector) == 2 * periods || error("power vector must have $(2 * periods) values")
    minimum = Float64.(definition["power_bounds_mw"]["minimum"])
    maximum = Float64.(definition["power_bounds_mw"]["maximum"])
    schedule = [zeros(Float64, periods), zeros(Float64, periods)]
    for plant in 1:2, period in 1:periods
        raw = clamp(vector[(plant - 1) * periods + period], 0.0, maximum[plant])
        schedule[plant][period] = raw < minimum[plant] / 2 ? 0.0 : max(raw, minimum[plant])
    end
    schedule
end

function evaluate!(worker::HydropowerWorker, vector::Vector{Float64}, definition; case_id="candidate")
    payload = Dict(
        "schema_version" => "synthetic-hydropower/v1",
        "requests" => [Dict(
            "case_id" => case_id,
            "incremental_inflow_m3s" => definition["incremental_inflow_m3s"],
            "target_power_mw" => project_power(vector, definition),
        )],
    )
    write(worker.io, JSON.json(payload), "\n")
    flush(worker.io)
    response = JSON.parse(readline(worker.io))
    haskey(response, "error") && error(response["error"]["message"])
    only(response["results"])
end

canonical_objective(worker::HydropowerWorker, vector::Vector{Float64}, definition) =
    Float64(evaluate!(worker, vector, definition)["simulation"]["objective"])
